### DISCUSSION TITLE
How can I use FastAPI in IIS?

### DISCUSSION THREAD

**Discussion by visualnetpr** [fastapi/fastapi#4207] [Questions] [answered]
Any way for FastAPI to work directly in IIS.

**Comment by igormorgado** [comment]
Really, are you a troll?

**Reply by visualnetpr** [reply]
Nop

**Comment by yezz123** [comment]
> Any way for FastAPI to work directly in IIS.

If we talk about the experience using IIS right now to deploy an Application is bad, and we see the developer experience is bad while using IIS, not like these new servers (Term of performance Linux servers are much better than Windows servers), but if want you could try to see if he support the ASGI Servers first and if uvicorn or starlette support the configuration of your IIS Server.

I use it once in a Django Project I found this repo https://github.com/mjkillough/iis-asgi-handler that's helped me a lot to handle the ASGI Server.
You can try to see if this can perform or have a similar server config, or maybe you could switch to azure at least, cause IIS how we see is dead and not a secure anymore service to use it

**Comment by INTzRct** [comment]
Yes !, You can. 
Because i already deploy FastAPI on IIS.

**Reply by visualnetpr** [reply]
How? Share the solution

**Reply by nmol2** [reply]
plz share the process

**Reply by RachataOutlier** [reply]
First of all please see this article 
[https://medium.com/nonstopio/deploy-django-application-on-windows-iis-server-93aee2864c41](url)
for know how to run Python on IIS

Step 2:
run command in your venv
`pip install a2wsgi`

and 
 
in main.py
`from fastapi import FastAPI
from a2wsgi import ASGIMiddleware

app = FastAPI()

@app.get("/")
def read_main():
    return {"message": "Hello World"}

wsgi_app = ASGIMiddleware(app)`

Step 3:
in web.config 
`<?xml version="1.0" encoding="UTF-8"?>
<configuration>
    <system.webServer>
        <handlers>
            <add name="FastAPI Handler" path="*" verb="*" modules="FastCgiModule" scriptProcessor="C:\inetpub\wwwroot\yourproject\yourvenv\Scripts\python.exe|C:\inetpub\wwwroot\yourproject\yourvenv\Lib\site-packages\wfastcgi.py" resourceType="Unspecified" />
        </handlers>
    </system.webServer>
	<appSettings>
  		<add key="PYTHONPATH" value="C:\inetpub\wwwroot\yourproject"/>
		<add key="WSGI_HANDLER" value="main.wsgi_app" />
	</appSettings>
</configuration>`

**Reply by visualnetpr** [reply]
One question, the link no open and said django not fastapi.

**Reply by INTzRct** [reply]
I will shown you to know how to config IIS for Python.
Do not need to create project for Django but you should make directory for FastAPI with main.py then change configuration follow my above comment.

**Reply by jusce17** [reply]
Thank @RachataOutlier ,
is there a way to extend  `--timeout-keep-alive`  ?

**Reply by KamalBalah** [reply]
> First of all please see this article [https://medium.com/nonstopio/deploy-django-application-on-windows-iis-server-93aee2864c41](url) for know how to run Python on IIS
> 
> Step 2: run command in your venv `pip install a2wsgi`
> 
> and
> 
> in main.py `from fastapi import FastAPI from a2wsgi import ASGIMiddleware
> 
> app = FastAPI()
> 
> @app.get("/") def read_main(): return {"message": "Hello World"}
> 
> wsgi_app = ASGIMiddleware(app)`
> 
> Step 3: in web.config `<?xml version="1.0" encoding="UTF-8"?> <configuration> <system.webServer> <handlers> <add name="FastAPI Handler" path="*" verb="*" modules="FastCgiModule" scriptProcessor="C:\inetpub\wwwroot\yourproject\yourvenv\Scripts\python.exe|C:\inetpub\wwwroot\yourproject\yourvenv\Lib\site-packages\wfastcgi.py" resourceType="Unspecified" /> </handlers> </system.webServer> <appSettings> <add key="PYTHONPATH" value="C:\inetpub\wwwroot\yourproject"/> <add key="WSGI_HANDLER" value="main.wsgi_app" /> </appSettings> </configuration>`

please where can i found web.config file

**Reply by KamalBalah** [reply]
> Yes !, You can. Because i already deploy FastAPI on IIS.

how to deploy on IIS please help

**Reply by jusce17** [reply]
@PowerRyder I followed these steps

**Reply by IgorSantRocha** [reply]
É só criar um no diretório do seu projeto:

C:\inetpub\wwwroot\TesteFastAPI\web.config

**Comment by nmol2** [comment, marked as answer]
I found these two articles that Helped me 

(https://docs.microsoft.com/en-us/visualstudio/python/configure-web-apps-for-iis-windows?view=vs-2019)
The first one recommending httpPlatformHandler in ISS for running Python Web Apps.

https://stackoverflow.com/questions/61919424/how-to-run-windows-iis-on-top-of-an-asgi-server-like-hypercorn-or-uvicorn
this second one shows how to use **httpPlatformHandler** for FastAPI specifically.

(there may be a chance that 'httpPlatformHandler' may not be present in your system, then download and install it via '**Microsoft Web Platform Installer**' available under Management Section on your IIS  screen of respective Website.

i have used hypercorn (alternative of unicorn suggested on the official fastapi website)
`pip install hypercorn`

web.config
```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
    <system.webServer>
     
        <handlers>
            <add name="httpPlatformHandler" path="*" verb="*" modules="httpPlatformHandler" resourceType="Unspecified" />
        </handlers>
        <httpPlatform processPath="C:\nmol\hpt-api\venv\Scripts\python.exe"
                    arguments="-m hypercorn app.main:app -b 127.0.0.1:%HTTP_PLATFORM_PORT% --keep-alive 5 --worker-class asyncio --workers 4"
                    stdoutLogEnabled="true" 
                    stdoutLogFile="C:\nmol\hpt-api\logs\python-stdout.log" 
                    startupTimeLimit="120" 
                    requestTimeout="00:05:00" 
                    startupRetryCount="3"
					processesPerApplication="5">
            <environmentVariables>
                <environmentVariable name="PORT" value="%HTTP_PLATFORM_PORT%" />
                
            </environmentVariables>
        </httpPlatform>
    </system.webServer>
</configuration>
```
(replace mentioned paths according to yours)

may this will work for you too

**Reply by jusce17** [reply]
I have replaced the paths as you mentioned but it keeps on saying : _The requested page cannot be accessed because the related configuration data for the page is invalid._

**Reply by PowerRyder** [reply]
@jusce17 Have you deployed it successfully? Can you please tell me the steps? I'm also trying to deploy FastApi to IIS but no success.

**Reply by KamalBalah** [reply]
> 

please can you explain the steps

**Reply by jusce17** [reply]
@KamalBalah , @PowerRyder 
I have jus added this to my `web.config `file 
```

<?xml version="1.0" encoding="UTF-8"?>
 <configuration>
  <system.webServer>

    <modules>
        <remove name="WebDAVModule" />
    </modules>
  	
   <handlers accessPolicy="Read, Execute, Script">
       
   <add name="PythonHandler" path="*" verb="*" modules="FastCgiModule" scriptProcessor="C:\MY-PROJECT-FOLDER\MY-PYTHON-ENV\Scripts\python.exe|C:\MY-PROJECT-FOLDER\MY-PYTHON-ENV\lib\site-packages\wfastcgi.py" resourceType="Unspecified" requireAccess="Script" />

	</handlers>

    <fastCgi>
<application fullPath="C:\MY-PROJECT-FOLDER\MY-PYTHON-ENV\Scripts\python.exe"  />
</fastCgi>

    <directoryBrowse enabled="true" />
        <tracing>
            <traceFailedRequests>
                <add path="*">
                    <traceAreas>
                        <add provider="ASP" verbosity="Verbose" />
                        <add provider="ASPNET" areas="Infrastructure,Module,Page,AppServices" verbosity="Verbose" />
                        <add provider="ISAPI Extension" verbosity="Verbose" />
                        <add provider="WWW Server" areas="Authentication,Security,Filter,StaticFile,CGI,Compression,Cache,RequestNotifications,Module,FastCGI,WebSocket,ANCM" verbosity="Verbose" />
                    </traceAreas>
                    <failureDefinitions statusCodes="500" />
                </add>
            </traceFailedRequests>
        </tracing>
 </system.webServer>
 <appSettings>
  <add key="PYTHONPATH" value="C:\MY-PROJECT-FOLDER" />
   <add key="WSGI_HANDLER" value="main.wsgi_app" /> 

</appSettings>
 </configuration> 
```


then you need to add this on your `main.py` file 

```
from a2wsgi import ASGIMiddleware
wsgi_app = ASGIMiddleware(app)
```

**Reply by Marv963** [reply]
@yichengsong318 do you found a Solution for deploying the app in iis?

**Reply by ghost** [reply]
please share that with the rest of us, thank you

**Reply by ghost** [reply]
Did you not declare `httpPlatform?` I am following this tutorial and it doesn't seem to have the same guidance as your example: https://learn.microsoft.com/en-us/visualstudio/python/configure-web-apps-for-iis-windows?view=vs-2022

**Reply by ghost** [reply]
I see. It works however this solution is based on a library that is no longer being maintained. I'll try to figure out how to use the httpPlatformHandler instead and update the ticket.

Thanks.

**Reply by aman-agrawal07** [reply]
hi yichengsong318,

i follow your guide now i get 404 httpcode do we need to start the uvicorn manually

**Reply by IgorSantRocha** [reply]
Consegui fazer subir usando o hypercorn também. Simples e rápido (já tinha o HttpPlatform instalado para outra aplicação, em Django).
Mas encontrei um problema, toda vez que eu mando uma requisição para o meu endpoint, são gerados 4 novos Workers.
Vou pesquisar para tentar encontrar uma forma de impedir que isso aconteça. Do contrario, vai travar meu servidor.

Enquanto isso, se alguém tiver alguma sugestão, agradeço!

**Reply by IgorSantRocha** [reply]
Resolvi mudando processesPerApplication="5" para "processesPerApplication="1".

Agora somente os 4 workers são gerados. Mas caso essa não seja uma boa solução. Adoraria saber!

**Reply by IgorSantRocha** [reply]
Sobre o erro 404, eu só consegui resolver ao criar um ambiente virtual e pedir para usar o python do ambiente:


processPath="C:\inetpub\wwwroot\MyAPP\venv\Scripts\python.exe"

**Reply by lextm** [reply]
The answer has several problems,

1. The steps are over-complicated and contain duplicate or useless settings.
2. No troubleshooting hints were provided.

I have a much simpler version on my blog https://docs.lextudio.com/blog/running-fastapi-web-apps-on-iis-with-httpplatformhandler/

**Reply by jstoppa** [reply]
I tried the steps suggested here but I get the error 
```
PermissionError: [WinError 10013] An attempt was made to access a socket in a way forbidden by its access permissions
```

Did anyone have this issue?

My web.config is the following
```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
    <system.webServer>
        <handlers accessPolicy="Read, Execute, Script">
            <remove name="httpPlatformHandler"/>
            <add name="httpPlatformHandler" path="*" verb="*" modules="httpPlatformHandler" resourceType="Unspecified" />
        </handlers>
        <httpPlatform processPath="C:\inetpub\wwwroot\FastAPI\Scripts\python.exe"
            arguments="-m hypercorn app.main:app --bind 127.0.0.1:8080 --keep-alive 5 --worker-class asyncio --workers 4"
            stdoutLogEnabled="true" stdoutLogFile="C:\logs\python.log" startupTimeLimit="120" requestTimeout="00:05:00"
            startupRetryCount="3" processesPerApplication="1">
        </httpPlatform>
        <httpErrors errorMode="Detailed" />
    </system.webServer>
</configuration>
```
If I stop IIS and I run the command below it works fine, it's only when running it inside IIS.
```bash
C:\inetpub\wwwroot\FastAPI\Scripts\python.exe -m hypercorn app.main:app --bind 127.0.0.1:8080 --keep-alive 5 --worker-class asyncio --workers 4

```
I also tried to grant IIS_IUSRS access to python.exe but it didn't make any difference, it was suggested in [this post](https://docs.lextudio.com/blog/running-flask-web-apps-on-iis-with-httpplatformhandler/#the-infinite-loading
) as when calling the API it hangs forever

I suspect it's something to do with the IIS port binding, if I change the web.config to point at another port (e.g. 8000) different than the binding port in IIS then I'm able to access the API (e.g. by navigating to http://localhost:8000), it feels to me IIS is using the port when hypercorn is trying to use it

**Reply by jstoppa** [reply]
I managed to get it working using uvicorn rather than hypercorn following the steps in this blog 
https://docs.lextudio.com/blog/running-fastapi-web-apps-on-iis-with-httpplatformhandler/

**Comment by oktaydbk54** [comment]
Hello, I wrote a code with FastAPI, I am trying to publish it on IIS, but my FastCGI option does not appear on the server, otherwise I would be very happy if you could help me to publish it.

**Reply by noahabe** [reply]
I don't know about deployment on IIS, but if you want to read more about deployment of a FastAPI project here you go: https://fastapi.tiangolo.com/deployment/
